util/types/datum.go

```diff
@@ -0,0 +1,519 @@
+	case KindUint64:
+		return d.compareUint64(ad.GetUint64())
+	case KindFloat64:
+		return d.compareFloat64(ad.GetFloat64())
+	case KindString:
+		return d.compareString(ad.GetString())
+	case KindBytes:
+		return d.compareBytes(ad.GetBytes())
+	case KindMysqlBit:
+		return d.compareMysqlBit(ad.GetMysqlBit())
+	case KindMysqlDecimal:
+		return d.compareMysqlDecimal(ad.GetMysqlDecimal())
+	case KindMysqlDuration:
+		return d.compareMysqlDuration(ad.GetMysqlDuration())
+	case KindMysqlEnum:
+		return d.compareMysqlEnum(ad.GetMysqlEnum())
+	case KindMysqlHex:
+		return d.compareMysqlHex(ad.GetMysqlHex())
+	case KindMysqlSet:
+		return d.compareMysqlSet(ad.GetMysqlSet())
+	case KindMysqlTime:
+		return d.compareMysqlTime(ad.GetMysqlTime())
+	case KindRow:
+		return d.compareRow(ad.GetRow())
+	default:
+		return 0
+	}
+}
+
+func (d *Datum) compareInt64(i int64) int {
+	switch d.k {
+	case KindInt64:
+		return CompareInt64(d.i, i)
+	case KindUint64:
+		if i < 0 || d.GetUint64() > math.MaxInt64 {
+			return 1
+		}
+		return CompareInt64(d.i, i)
+	default:
+		return d.compareFloat64(float64(i))
+	}
+}
+
+func (d *Datum) compareUint64(u uint64) int {
+	switch d.k {
+	case KindInt64:
+		if d.i < 0 || u > math.MaxInt64 {
+			return -1
+		}
+		return CompareInt64(d.i, int64(u))
+	case KindUint64:
+		return CompareUint64(d.GetUint64(), u)
+	default:
+		return d.compareFloat64(float64(u))
+	}
+}
+
+func (d *Datum) compareFloat64(f float64) int {
+	switch d.k {
+	case KindNull:
+		return -1
+	case KindInt64:
+		return CompareFloat64(float64(d.i), f)
+	case KindUint64:
+		return CompareFloat64(float64(d.GetUint64()), f)
+	case KindFloat64:
+		return CompareFloat64(d.GetFloat64(), f)
+	case KindString, KindBytes:
+		fVal := parseFloatLoose(d.GetString())
+		return CompareFloat64(fVal, f)
+	case KindMysqlBit:
+		fVal := d.GetMysqlBit().ToNumber()
+		return CompareFloat64(fVal, f)
+	case KindMysqlDecimal:
+		fVal, _ := d.GetMysqlDecimal().Float64()
+		return CompareFloat64(fVal, f)
+	case KindMysqlDuration:
+		fVal := d.GetMysqlDuration().Seconds()
+		return CompareFloat64(fVal, f)
+	case KindMysqlEnum:
+		fVal := d.GetMysqlEnum().ToNumber()
+		return CompareFloat64(fVal, f)
+	case KindMysqlHex:
+		fVal := d.GetMysqlHex().ToNumber()
+		return CompareFloat64(fVal, f)
+	case KindMysqlSet:
+		fVal := d.GetMysqlSet().ToNumber()
+		return CompareFloat64(fVal, f)
+	case KindMysqlTime:
+		fVal, _ := d.GetMysqlTime().ToNumber().Float64()
+		return CompareFloat64(fVal, f)
+	default:
+		return -1
+	}
+}
+
+func numPart(s string) string {
+	s = strings.TrimSpace(s)
+	var hasDot bool
+	var i int
+	for i = 0; i < len(s); i++ {
+		if s[i] == '.' {
+			if hasDot {
+				break
+			}
+			hasDot = true
+			continue
+		}
+		if s[i] < '0' && s[i] > '9' {
+			break
+		}
+	}
+	if i == 0 {
+		return "0"
+	}
+	return s[:i]
+}
+
+func parseFloatLoose(s string) float64 {
+	s = numPart(s)
+	val, err := strconv.ParseFloat(s, 64)
+	if err != nil {
+		log.Errorf("should not happen!")
```

Comment:
mysql> select 1.5 > "xxx";
+-------------+
| 1.5 > "xxx" |
+-------------+
|           1 |
+-------------+
1 row in set, 1 warning (0.01 sec)